modules/index/src/main/java/org/apache/ignite/internal/idx/InternalSortedIndexImpl.java

```diff
@@ -0,0 +1,270 @@
+    public InternalSortedIndexImpl(IgniteUuid id, String name, SortedIndexStorage store, TableImpl tbl) {
+        this.id = id;
+        this.name = name;
+        this.store = store;
+        this.tbl = tbl;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public IgniteUuid id() {
+        return id;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String name() {
+        return name;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String tableName() {
+        return tbl.name();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public List<Column> columns() {
+        return store.indexDescriptor().columns().stream()
+                .map(SortedIndexColumnDescriptor::column)
+                .collect(Collectors.toList());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Cursor<Tuple> scan(Tuple low, Tuple up, byte scanBoundMask, BitSet proj) {
+        Cursor<IndexRow> cur = store.range(
+                low != null ? new IndexSearchRow(low) : null,
+                up != null ? new IndexSearchRow(up) : null,
+                r -> true
+        );
+
+        List<String> tblColIds = new ArrayList<>(tbl.schemaView().schema().columnNames());
+        Set<String> idxColIds = columns().stream().map(Column::name).collect(Collectors.toSet());
+
+        boolean needLookup = proj.stream().anyMatch(order -> !idxColIds.contains(tblColIds.get(order)));
+
+        return new IndexCursor(cur, needLookup ? this::convertWithTableLookup : this::convertIndexedOnly);
+    }
+
+    @Override
+    public void drop() {
+        store.destroy();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onUpdate(@Nullable BinaryRow oldRow, BinaryRow newRow, int partId) {
+        Tuple t = TableRow.tuple(tbl.schemaView().resolve(newRow));
+
+        IndexBinaryRow idxBinRow = store.indexRowFactory().createIndexRow(t, newRow.keyRow(), partId);
+
+        store.put(idxBinRow);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onRemove(BinaryRow row) {
+        Tuple t = TableRow.tuple(tbl.schemaView().resolve(row));
+
+        IndexBinaryRow idxBinRow = store.indexRowFactory().createIndexRow(t, row.keyRow(), -1);
+
+        store.remove(idxBinRow);
+    }
+
+    /**
+     * Used for index only scan.
+     */
+    private Tuple convertIndexedOnly(IndexRow r) {
+        assert r.columnsCount() == store.indexDescriptor().columns().size() : "Unexpected Index row to convert [idxRow=" + r
+                + ", index=" + this + ']';
+
+        Tuple t = Tuple.create();
+
+        // Create the tuple with column order similar to table Binary Row column order.
+        Arrays.stream(tbl.schemaView().schema().keyColumns().columns())
+                .map(Column::name)
+                .forEach(colName -> t.set(colName, null));
+
+        Arrays.stream(tbl.schemaView().schema().valueColumns().columns())
+                .map(Column::name)
+                .forEach(colName -> t.set(colName, null));
```

Comment:
Actually, Tuple doesn't imply any column order.
What is the purpose of these lines?

Reply:
> Tuple doesn't imply any column order.

What do you mean?
The method `Tuple#value(int)` explicitly use column index (in other words column order).